### DISCUSSION TITLE
How t get user Permission in the user Profile using a custom SPI

### DISCUSSION THREAD

**Discussion by taru-j-a-i-n** [keycloak/keycloak#29187] [Q&A]
### Description

I want to see the User permission in the user profile of the user but there is no direct method to do so please provide the direct method or the solution to get User permissions .
the output must have the  Pwrmissions as not null and not empty currently it is coming as an empty list as shown below: {"id":"094073fb-0f5e-419f-a582-d25b8b5b5903","email":"admin@admin.com","firstName":"Admin","lastName":"listener","userRoles":["default-roles-portal-realm","create-client","uma_protection","read-token","view-profile","view-groups","view-consent","view-applications","manage-consent","manage-account-links","delete-account"],"userPermissions":[],"claims":null}

**I tried this method but it is not giving the solution: ** 

public List<UserPermission> getPermissions(UserModel user) {

        String realmName = System.getenv("realm_name");
        RealmModel realm = this.model.getRealmByName(realmName);

        ClientModel client = realm.getClientByClientId("demo-client");
        log.info("client: " + client);

        log.info("getPermissions is triggered...........................");
        List<UserPermission> userPermissions = new ArrayList<>();
        StoreFactory storeFactory = session.getProvider(StoreFactory.class);
        log.info("storeFactory: " + storeFactory);

        AuthorizationProvider authorizationProvider = session.getProvider(AuthorizationProvider.class);
         log.info("authorizationProvider: " + authorizationProvider);
         
        PermissionTicketStore permissionTicketStore = storeFactory.getPermissionTicketStore();
        log.info("permissionTicketStore: " + permissionTicketStore);
        ResourceServerStore resourceServerStore = storeFactory.getResourceServerStore();
        log.info("resourceServerStore: " + resourceServerStore);
        ResourceServer resourceServer = resourceServerStore.findByClient(client);

        log.info("resourceServer: " + resourceServer);
        if (resourceServer != null) {
            log.info("resourceServer is not null...........................");
            List<PermissionTicket> permissionTickets = permissionTicketStore.findGranted(resourceServer, user.getId());
            log.info("user.getId(): " + user.getId());
            log.info("permissionTickets: " + permissionTickets);
            for (PermissionTicket permissionTicket : permissionTickets) {
                log.info("permissionTicket: loop:  " + permissionTicket);
                UserPermission userPermission = new UserPermission();
                userPermission.setResourceId(permissionTicket.getResource().getId());
                log.info("permissionTicket.getResource().getId(): " + permissionTicket.getResource().getId());
                userPermission.setResourceName(permissionTicket.getResource().getName());
                log.info("permissionTicket.getResource().getName(): " + permissionTicket.getResource().getName());
                userPermission.setActions(getScopeNames(permissionTicket.getResource()));
                log.info("getScopeNames(permissionTicket.getResource()): " + getScopeNames(permissionTicket.getResource()));
                userPermissions.add(userPermission);
            }
        }

        return userPermissions;
    }
```[tasklist]
### Tasks
```

**Comment by codespearhead** [comment]
> I want to see the User permission in the user profile of the user

Do you want any user to see what permissions they have? If so, what are you trying to accomplish with this custom SPI? 

Because "retrieving user permissions" doesn't make much sense unless you're strictly using [role-based policies](https://www.keycloak.org/docs/latest/authorization_services/#_policy_rbac): just imagine a resource you can only accessed at a certain time of the day.

**Reply by taru-j-a-i-n** [reply]
I am **using Role based policy permission only** if the user have a particular role suppose he is a github admin then whatever policy permission assigned to him should come under the permission I don’t know whether it makes sense or not i want to see all the permissions align with a particular user in the user profile whether by using policy evaluator or By permission ticiet or by evaluator or by whatever means you can provide me 2 solutions one for providing permissions normally one for role based policies if u want.
